src/sql/workbench/contrib/notebook/browser/notebookLinkHandler.ts

```diff
@@ -0,0 +1,145 @@
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { URI } from 'vs/base/common/uri';
+import * as path from 'vs/base/common/path';
+import { IConfigurationService } from 'vs/platform/configuration/common/configuration';
+import { replaceInvalidLinkPath } from 'sql/workbench/contrib/notebook/common/utils';
+import { isWindows } from 'vs/base/common/platform';
+
+const keepAbsolutePathConfigName = 'notebook.keepAbsolutePath';
+
+export class NotebookLinkHandler {
+	private _notebookUriLink: URI;
+	private _href: string;
+	private _notebookDirectory: string;
+	private _isAnchorLink: boolean;
+	private _isFile: boolean;
+	public readonly isAbsolutePath: boolean;
+
+	constructor(
+		private _notebookURI: URI,
+		private _link: string | HTMLAnchorElement,
+		@IConfigurationService private _configurationService: IConfigurationService,
+	) {
+		if (typeof this._link === 'string') {
+			this._notebookUriLink = URI.parse(this._link);
+			this._isFile = this._notebookUriLink.scheme === 'file';
+			this.isAbsolutePath = path.isAbsolute(this._link);
+			this._isAnchorLink = this._link.includes('#') && this._isFile;
+		} else {
+			// HTMLAnchorElement
+			// windows files need to use the link.href instead as it contains the file:// prefix
+			// which enables us to get the proper relative path
+			if (isWindows) {
+				this._href = this._link.href;
+			} else {
+				this._href = this._link.attributes['href']?.nodeValue;
+			}
+			this._notebookUriLink = this._href ? URI.parse(this._href) : undefined;
+			this._isFile = this._link.protocol === 'file:';
+			this._isAnchorLink = this._notebookUriLink?.fragment ? true : false;
+			this.isAbsolutePath = this._link.attributes['is-absolute']?.nodeValue === 'true' ? true : false;
+		}
+		this._notebookDirectory = this._notebookURI ? path.dirname(this._notebookURI.fsPath) : '';
+	}
+
+	/**
+	 * Function to get the link for LinkCalloutDialog or htmlMarkdownConverter
+	 * When a user inserts a new link via the LinkCalloutDialog it will go through the string case to
+	 * get the absolute path of the file and then will be converted to anchor element that will be called again
+	 * to the object case in which we will find the relative path of the file unless the user has the
+	 * keep absolute setting enabled then we don't convert absolute paths to relative paths
+	 * @returns the file link or web link
+	 */
+	public getLinkUrl(): string {
+		// cases where we only have the href link
+		if (typeof this._link === 'string') {
+			// Does not convert absolute path to relative path
+			if (this._isFile && this.isAbsolutePath && this._configurationService.getValue(keepAbsolutePathConfigName) === true) {
+				return this._link;
```

Comment:
I noticed that by adding an absolute path with spaces and the keepAbsolute setting set to true the markdown link is not constructed properly. However, I do not see this issue when the keepAbsolute setting is set to false, because the relative link adds the %20 instead. 
`[/Users/{user}/hybrid book/content/readme.md](/Users/{user}/hybrid book/content/readme.md)`

Reply:
Fixed by escaping the href when the setting is on but will fix the other %20 issues in the other PR #16422.